opentelemetry-api/opentelemetry/backend.py

```diff
@@ -0,0 +1,131 @@
+_UNIT_TEST_IGNORE_ENV = False
+
+_tracer = None
+
+def _get_fallback_impl(api_type: Type[_T]) -> _T:
+    # TODO: Move (most of) this to module docstring.
+    """Gets the fallback implementation for ``api_type``.
+
+    ``api_type`` must be a OpenTelemetry API type like :cls:`Tracer`.
+
+    First, the function tries to find a module that provides a `get_opentelemetry_backend_impl`
+    function (with the same signature as this function). The following modules are tried:
+
+    1. `$OPENTELEMETRY_PYTHON_BACKEND_<typename>` (e.g. `OPENTELEMETRY_PYTHON_BACKEND_TRACER`)
+    2. `$OPENTELEMETRY_PYTHON_BACKEND_DEFAULT` (e.g. `OPENTELEMETRY_PYTHON_BACKEND_TRACER`)
+    3. The OpenTelemetry SDK's tracer module.
+
+    Note that if any of the environment variables is set to an nonempty value, further steps
+    are not tried, even if the modulename set there is invalid or fails to load. The no-op API
+    implementation is returned instead.
+    """
+    backend_modname = None
+    if not _UNIT_TEST_IGNORE_ENV and not sys.flags.ignore_environment:
+        backend_modname = os.getenv('OPENTELEMETRY_PYTHON_BACKEND_' + api_type.__name__.upper())
+        if not backend_modname:
+            backend_modname = os.getenv('OPENTELEMETRY_PYTHON_BACKEND_DEFAULT')
+    if not backend_modname:
+        backend_modname = _DEFAULT_BACKEND_MODNAME
+    try:
+        backend_mod = importlib.import_module(backend_modname)
+    except (ImportError, SyntaxError):
+        # TODO Log/warn
+        return api_type()
+    try:
+        # Note: We use such a long name to avoid calling a function that is not intended for this
+        # API.
+        backend_fn: Callable[[Type[_T]], object] = getattr(backend_mod, 'get_opentelemetry_backend_impl')
+    except AttributeError:
+        # TODO Log/warn
+        return api_type()
+    result = backend_fn(api_type)
+    if result and isinstance(result, api_type):
+        # TODO: Warn if backend_fn returns api_type(): It should return None to indicate using the
+        # default.
+        return result
+    return api_type()
+
+def _selectimpl(api_type: Type[_T]) -> _T:
+    impl = _get_fallback_impl(api_type)
+    assert impl
+    globals()[api_type.__name__.lower()] = impl
+    return impl
+
+def _set_backend_object(api_type: Type[_T], impl_object: _T) -> None:
+    if impl_object is None:
+        raise ValueError('None is not allowed as a backend implementation.')
+    if not isinstance(impl_object, api_type):
+        raise ValueError('The object does not implement the required base class.')
+    globals()['_' + api_type.__name__.lower()] = impl_object
+
+### Public code (basically copy & paste for each type) {{{1
+
+def tracer() -> Tracer:
+    """Gets the current global :cls:`Tracer` object.
+
+    If there isn't one set yet, a default will be used (see module documentation).
+    """
+
+    if _tracer:
+        return _tracer
+    return _selectimpl(Tracer)
+
+def set_tracer(tracer_implementation: Tracer) -> None:
```

Comment:
I'm fine with disallowing that after a Tracer has been initially selected. It is mostly there to explicitly set a particular instance (maybe configured in an app-specific way) of the Tracer.

Reply:
I now only provide a setter for the factory function.

Reply:
Actually I like offering both a `Tracer` and a factory, as long as they cannot be re-set.

Another thing is that usually code has higher priority than environment variables (hence, if the user were to use this function explicitly, it would have precedence over the rest). Opinions on this?